Source/CountDown.spoon/init.lua

```diff
@@ -43,45 +133,145 @@ end
+   obj.currentIcon = obj.iconActive
+   obj.loopCount = 0
+   
+   if mainRes.w/secCount >= minimumStep then
+      -- do every second
+      obj.timerBar = hs.timer.doEvery(1, function()
+                                         obj:setProgress(obj.progress)
+      end)
+   else
+      -- we need few seconds to advance two pixels
+      local interval = 2/(mainRes.w/secCount)
+      obj.timerBar = hs.timer.doEvery(interval,
+                                      function()
+                                         obj:setProgress(obj.progress)
+                                      end
+      )
+   end
+
+end
+
+--- CountDown:time_absolute_seconds()
+--- Method
+---    Return current time in seconds
+---
+--- Parameters:
+---  none
+function obj:time_absolute_seconds()
+   -- return hs.timer.absoluteTime() in seconds
+   local timeNow = hs.timer.absoluteTime()
+   return math.floor(timeNow / 1e9)
+end
+
+--- CountDown:tick()
+--- Method
+---    Callback executed every second
+---
+--- Parameters:
+---  none
+function obj:tick()
+   -- ticks every second...
+   local timeNow = obj:time_absolute_seconds()
+   obj.timeLeft = (obj.startTime + obj.lenMin * 60) - timeNow
+   if obj.timeLeft <= 0 then
+      obj.timeLeft = 0
+      obj:reset()
+      obj:notify()
+      obj.lenMin = nil
+      obj.progress = 0
+   else --update progress ratio
+      obj.progress = 1 - obj.timeLeft/(obj.lenMin * 60)
+   end
+end
+
+function obj:create_time_timer(minutes)
+   obj.timerRunning = true
+   obj.lenMin = minutes
+   obj.timeLeft = minutes * 60
+   obj.startTime = obj:time_absolute_seconds()
+   obj.timer = hs.timer.doWhile(function() return obj.timerRunning end,
+      function() obj:tick() end,
+      1)
+end
+
+--- CountDown:startFor(minutes)
+--- Method
+---    Start the time for the specified time
+---
+--- Parameters:
+---    minutes: number of minutes the timer will run
+---
 function obj:startFor(minutes)
-    if obj.timer then
-        canvasCleanup()
-    else
-        local mainScreen = hs.screen.mainScreen()
-        local mainRes = mainScreen:fullFrame()
-        obj.canvas:frame({x=mainRes.x, y=mainRes.h-5, w=mainRes.w, h=5})
-        -- Set minimum visual step to 2px (i.e. Make sure every trigger updates 2px on screen at least.)
-        local minimumStep = 2
-        local secCount = math.ceil(60*minutes)
-        obj.loopCount = 0
-        if mainRes.w/secCount >= 2 then
-            obj.timer = hs.timer.doEvery(1, function()
-                obj.loopCount = obj.loopCount+1/secCount
-                obj:setProgress(obj.loopCount, minutes)
-            end)
-        else
-            local interval = 2/(mainRes.w/secCount)
-            obj.timer = hs.timer.doEvery(interval, function()
-                obj.loopCount = obj.loopCount+1/mainRes.w*2
-                obj:setProgress(obj.loopCount, minutes)
-            end)
-        end
+   if obj.timer then
+      -- reset current timer and do nothing
+      obj:reset()
+   else
```

Comment:
what if the `minutes` value is changed?

I think it is better to cleanup if there a timer exists - or check if minutes are change
```suggestion
   if obj.timer then
      -- reset current timer and do nothing
      obj:reset()
   end
```